### DISCUSSION TITLE
Can conflicting/related mutations be consolidated?

### DISCUSSION THREAD

**Discussion by PabloRomanH** [TanStack/query#5083] [Q&A]
I'm adopting an offline-first approach on my website with optimistic updates everywhere where it's possible.

I'm finding that when offline, many mutations can be created that could be reduced to a much smaller number of mutations, because they are just updating the same value. I'm finding two patterns of types of mutations that would be great to be able to join into fewer calls to the backend:

- Consecutive mutations that update the same value. In this case, only the last one should be executed.
- Cumulative mutations. Each mutation adds a certain amount to a value. By adding up the amounts in all the mutations that are waiting to be executed, I can calculate a total amount to be added, and replace all of them with a single mutation.

Is this possible?

**Comment by TkDodo** [comment]
It's an interesting request. Currently, we just execute all paused mutations, one after the other, via `resumePausedMutations()`:

https://github.com/TanStack/query/blob/aa94bdd5b67eefb5065834645f2dadc715e85812/packages/query-core/src/mutationCache.ts#L163-L180

I thought about adding a `muationFilter` to that function, and while this would work (I think), the other mutations would still remain in the cache in paused state and might be resumed at another time. Also, _we_ call that function, not you :/

So I think what you'd have to do is hook into the `mutationCache` and delete / consolidate mutations on your own. I think you could `subscribe` to the `mutationCache`, listen to events and then manipulate the cache, e.g. by calling `remove` on older mutations. You might have to tag them with a `muationKey` so that you know which ones belong together ...
